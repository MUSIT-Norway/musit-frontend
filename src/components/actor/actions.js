import { Observable, Subject } from 'rxjs/Rx';
import * as loglevel from 'loglevel';
import 'whatwg-fetch';
import { getToken } from '../../middleware/ApiClient'

// Reducer subjects (for data)
export const update$ = new Subject();
export const clear$ = new Subject();

// Async actions
export const input$ = new Subject();
input$.map(update => update.value)
  .filter((text) => text.length > 2)
  .debounce(() => Observable.timer(700))
  .distinctUntilChanged()
  .switchMap((term) =>
    fetch(`/api/actor/v1/person?search=[${term}]&museumId=1`, {
      headers: {
        Authorization: `Bearer ${getToken()}`
      }
    }).then((response) => response.json())
  )
  .subscribe(
    (data) => update$.next(data),
    (error) => {
      loglevel.error('Error:' + error);
      clear$.next();
    }
  );