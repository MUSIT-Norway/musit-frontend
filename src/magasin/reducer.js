import { toFrontend } from './mapper';
import * as types from './types';

const initialState = {
  root: {
    data: {}
  }
};

const storageUnitGridReducer = (state = initialState, action = {}) => {
  switch (action.type) {
  case types.LOAD_SEVERAL:
    return {
      ...state,
      loading: true
    };
  case types.LOAD_SEVERAL_SUCCESS:
    return {
      ...state,
      loading: false,
      loaded: true,
      data: action.result
    };
  case types.LOAD_SEVERAL_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      error: action.error
    };
  case types.LOAD_ONE_SUCCESS:
    return {
      ...state,
      root: {
        ...state.root,
        loading: false,
        loaded: true,
        data: toFrontend(action.result)
      }
    };
  case types.LOAD_ONE_FAIL:
    return {
      ...state,
      root: {
        ...state.root,
        loading: false,
        loaded: false,
        error: action.error
      }
    };
  case types.CLEAR_ROOT: {
    return {
      ...state,
      root: {}
    };
  }
  case types.DELETE:
    return {
      ...state,
      loading: true
    };
  case types.DELETE_SUCCESS:
    return {
      ...state,
      loading: false,
      loaded: true,
      data: []
    };
  case types.DELETE_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      error: action.error
    };
  default:
    return state;
  }
};

export default storageUnitGridReducer;