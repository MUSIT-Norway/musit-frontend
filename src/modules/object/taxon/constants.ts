export const stadiumList = [
  {
    code: '',
    term: '--select--'
  },
  {
    code: 'e',
    term: 'egg'
  },
  {
    code: 'l',
    term: 'larvae'
  },
  {
    code: 'p',
    term: 'pupae'
  },
  {
    code: 'x',
    term: 'pupae exuviae'
  },
  {
    code: 'i',
    term: 'imago'
  },
  {
    code: 's',
    term: 'subimago'
  },
  {
    code: 'j',
    term: 'juvenile'
  },
  {
    code: 'n',
    term: 'nymph'
  },
  {
    code: 'a',
    term: 'adult'
  },
  {
    code: 'u',
    term: 'unknown'
  }
];

export const sexList = [
  {
    code: '',
    term: '--select--'
  },
  {
    code: 'm',
    term: 'male'
  },
  {
    code: 'f',
    term: 'female'
  },
  {
    code: 'h',
    term: 'hermaphrodite'
  },
  {
    code: 'd',
    term: 'doubtful gender'
  },
  {
    code: 'u',
    term: 'unknown'
  }
];
