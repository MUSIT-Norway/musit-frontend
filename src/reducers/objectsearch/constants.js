export const SEARCH_OBJECTS = 'musit/objectSearch/SEARCH'
export const SEARCH_OBJECTS_SUCCESS = 'musit/objectSearch/SEARCH_SUCCESS'
export const SEARCH_OBJECTS_FAIL = 'musit/objectSearch/SEARCH_FAIL'
export const CHANGE_FIELD = 'musit/objectSearch/CHANGE_FIELD'