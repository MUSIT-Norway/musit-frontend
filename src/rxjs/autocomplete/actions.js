import { Subject } from 'rxjs/Rx';
import { Observable } from 'rxjs/Rx';
import * as loglevel from 'loglevel';
import 'whatwg-fetch';
import App from '../../modules/app/index';
const { getAccessToken, getMuseumId } = App.actions;

export default function createActions(urlFun) {
  // Reducer subjects (for data)
  const update$ = new Subject();
  const clear$ = new Subject();

  // Async actions
  const input$ = new Subject();
  input$
    .debounce(() => Observable.timer(500))
    .distinctUntilChanged()
    .switchMap((term) =>
      fetch(urlFun(term, getMuseumId()), {
        headers: {
          Authorization: `Bearer ${getAccessToken()}`
        }
      }).then((response) => response.json())
    )
    .subscribe(
      (data) => update$.next(data),
      (error) => {
        loglevel.error('Error:' + error);
        clear$.next();
      }
    );

  return {
    input$,
    update$,
    clear$
  };
}

