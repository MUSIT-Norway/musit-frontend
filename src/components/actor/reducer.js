import { Observable } from 'rxjs/Rx';
import actions from "./actions";

const dataFromServer = [
  {
    id: 1,
    fn: 'Jarl parl'
  },
  {
    id: 2,
    fn: 'Line pine'
  },
  {
    id: 3,
    fn: 'Kjell pell'
  },
  {
    id: 4,
    fn: 'Stein sein'
  },
  {
    id: 5,
    fn: 'Knut lut'
  },
  {
    id: 6,
    fn: 'Ellen'
  }
]

const initialState = [];

const ActorReducer$ = Observable.of(() => initialState)
  .merge(
    actions.suggest$.map((update) => () => {
      if (!update) {
        return initialState;
      }
      return dataFromServer.filter(d => {
        const name = d.fn.toLowerCase();
        const query = update.value.toLowerCase();
        return name.indexOf(query) > -1
      })
    }),
    actions.clear$.map(() => () => initialState),
  );

export default Observable.merge(
  ActorReducer$.map(reducer => ["suggest", reducer]),
);