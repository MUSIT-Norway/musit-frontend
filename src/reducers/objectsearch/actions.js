import * as types from './constants'
import { apiUrl } from'../../util'
import Config from '../../config'

export const searchForObjects = (params, page) => {
  const baseUrl = apiUrl(Config.magasin.urls.objectsearch.baseUrl(1));
  const museumNoQuery = `museumNo=${params.museumNo || ''}`
  const subNoQuery = `subNo=${params.subNo || ''}`
  const termQuery = `term=${params.term || ''}`
  const pageQuery = `page=${page}`
  const limitQuery = `limit=${params.perPage}`
  const url = `${baseUrl}?${museumNoQuery}&${subNoQuery}&${termQuery}&${pageQuery}&${limitQuery}`
  return {
    types: [types.SEARCH_OBJECTS, types.SEARCH_OBJECTS_SUCCESS, types.SEARCH_OBJECTS_FAIL],
    promise: (client) => client.get(url),
    page
  }
}

export const onChangeField = (field, value) => {
  return {
    type: types.CHANGE_FIELD,
    field,
    value
  }
}