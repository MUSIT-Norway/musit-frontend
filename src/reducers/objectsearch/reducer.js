import { omit } from 'lodash';
import { getPath } from '../helper';
import MusitObject from '../../models/object';
import * as types from './constants';

const initialState = {
  data: {
    totalMatches: 0,
    matches: []
  },
  params: {
    currentPage: 1,
    perPage: 50
  },
  loaded: true
};

export default (state = initialState, action) => {
  switch(action.type) {
  case types.CHANGE_FIELD:
    return {
      ...state,
      params: {
        ...state.params,
        [action.field]: action.value
      }
    };
  case types.SEARCH_OBJECTS:
    return {
      ...omit(state, 'error'),
      loading: true,
      loaded: false,
      data: {
        totalMatches: 0,
        matches: []
      },
      params: {
        ...state.params,
        currentPage: action.page
      }
    };
  case types.SEARCH_OBJECTS_FAIL:
    return {
      ...state,
      loading: false,
      loaded: false,
      data: [],
      error: action.error
    };
  case types.SEARCH_OBJECTS_SUCCESS:
    return {
      ...omit(state, 'error'),
      loading: false,
      loaded: true,
      data: {
        totalMatches: action.result.totalMatches,
        matches: action.result.matches.map(data => {
          return new MusitObject({
            ...data,
            breadcrumb: getPath(data)
          });
        })
      }
    };
  default:
    return state;
  }
};