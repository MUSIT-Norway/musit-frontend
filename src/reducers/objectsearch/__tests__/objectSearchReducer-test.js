import * as actions from '../actions'
import * as constants from '../constants'
import reducer from '../reducer'
import Config from '../../../config'
import request from 'superagent';
import nocker from 'superagent-nock';
const nock = nocker(request);

describe('Object search reducer', () => {
    it('creates SEARCH_OBJECTS_SUCCESS when fetching data has been done', () => {
        const params = {
            museumNo:'',
            subNo: '',
            term: ''
        }
        const page = {
            page: 1,
            perPage: 50
        }
        const baseUrl = Config.magasin.urls.objectsearch.baseUrl(1);
        const museumNoQuery = `museumNo=${params.museumNo || ''}`
        const subNoQuery = `subNo=${params.subNo || ''}`
        const termQuery = `term=${params.term || ''}`
        const pageQuery = `page=${page}`
        const limitQuery = `limit=${params.perPage}`
        const url = `${baseUrl}?${museumNoQuery}&${subNoQuery}&${termQuery}&${pageQuery}&${limitQuery}`
        nock('http://localhost')
            .get(url)
            .reply(200, {
                totalMatches: 1,
                matches:
                    [
                        {
                            id: 51,
                            museumNo: 'MusN11',
                            term: 'Solsikke'
                        }
                        ]
            })
        const store = mockStore()

        return store.dispatch(actions.searchForObjects(params, page))
            .then(() => {
                expect(store.getActions()).toMatchSnapshot()
            })
    })

    it('search objects: no action', () => {
        expect(
            reducer(undefined, {})
        ).toMatchSnapshot()
    })

    it('search objects: initial action', () => {
        expect(
            reducer(undefined, {
                type: constants.SEARCH_OBJECTS
            })
        ).toMatchSnapshot()
    })

    it('search objects: success action', () => {
        expect(
            reducer(undefined, {
                type: constants.SEARCH_OBJECTS_SUCCESS,
                result: {
                    someField: 1
                }
            })
        ).toMatchSnapshot()
    })

    it('search objects: fail action', () => {
        expect(
            reducer(undefined, {
                type: constants.SEARCH_OBJECTS_FAIL,
                error: Error('search objects has some error.')
            })
        ).toMatchSnapshot()
    })
})
  