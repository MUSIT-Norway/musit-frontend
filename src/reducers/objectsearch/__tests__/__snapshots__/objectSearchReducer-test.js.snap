exports[`Object search reducer creates SEARCH_OBJECTS_SUCCESS when fetching data has been done 1`] = `
Array [
  Object {
    "page": Object {
      "page": 1,
      "perPage": 50
    },
    "type": "musit/objectSearch/SEARCH"
  },
  Object {
    "page": Object {
      "page": 1,
      "perPage": 50
    },
    "result": Object {
      "matches": Array [
        Object {
          "id": 51,
          "museumNo": "MusN11",
          "term": "Solsikke"
        }
      ],
      "totalMatches": 1
    },
    "type": "musit/objectSearch/SEARCH_SUCCESS"
  }
]
`;

exports[`Object search reducer search objects: fail action 1`] = `
Object {
  "data": Array [],
  "error": [Error: search objects has some error.],
  "loaded": false,
  "loading": false,
  "params": Object {
    "currentPage": 1,
    "perPage": 50
  }
}
`;

exports[`Object search reducer search objects: initial action 1`] = `
Object {
  "data": Object {
    "matches": Array [],
    "totalMatches": 0
  },
  "loaded": false,
  "loading": true,
  "params": Object {
    "currentPage": undefined,
    "perPage": 50
  }
}
`;

exports[`Object search reducer search objects: no action 1`] = `
Object {
  "data": Object {
    "matches": Array [],
    "totalMatches": 0
  },
  "params": Object {
    "currentPage": 1,
    "perPage": 50
  }
}
`;

exports[`Object search reducer search objects: success action 1`] = `
Object {
  "data": Object {
    "matches": Array [
      MusitObject {
        "breadcrumb": Array [
          Object {
            "id": 1,
            "name": "root-node",
            "url": "/magasin/1"
          },
          Object {
            "id": 2,
            "name": "Utviklingsmuseet",
            "url": "/magasin/2"
          },
          Object {
            "id": 3,
            "name": "Forskningens hus",
            "url": "/magasin/3"
          },
          Object {
            "id": 4,
            "name": "Kulturværelset",
            "url": "/magasin/4"
          },
          Object {
            "id": 5,
            "name": "Naturværelset",
            "url": "/magasin/5"
          }
        ],
        "currentLocationId": 5,
        "id": 51,
        "mainObjectId": undefined,
        "museumNo": "MusN11",
        "pathNames": Array [
          Object {
            "name": "root-node",
            "nodeId": 1
          },
          Object {
            "name": "Utviklingsmuseet",
            "nodeId": 2
          },
          Object {
            "name": "Forskningens hus",
            "nodeId": 3
          },
          Object {
            "name": "Kulturværelset",
            "nodeId": 4
          },
          Object {
            "name": "Naturværelset",
            "nodeId": 5
          }
        ],
        "subNo": undefined,
        "term": "Solsikke"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 54,
        "mainObjectId": undefined,
        "museumNo": "MusN13",
        "pathNames": undefined,
        "subNo": undefined,
        "term": "Makrellsopp"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 19,
        "mainObjectId": undefined,
        "museumNo": "MusN20",
        "pathNames": undefined,
        "subNo": undefined,
        "term": "Ukjent mygg"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 17,
        "mainObjectId": undefined,
        "museumNo": "MusN21",
        "pathNames": undefined,
        "subNo": "2",
        "term": "Snilebille"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 18,
        "mainObjectId": undefined,
        "museumNo": "MusN22",
        "pathNames": undefined,
        "subNo": undefined,
        "term": "Fluesopp"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 14,
        "mainObjectId": undefined,
        "museumNo": "MusK23",
        "pathNames": undefined,
        "subNo": undefined,
        "term": "Lite skaft av ben"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 15,
        "mainObjectId": undefined,
        "museumNo": "MusK24",
        "pathNames": undefined,
        "subNo": "a",
        "term": "Lendeklede"
      },
      MusitObject {
        "breadcrumb": Array [],
        "currentLocationId": undefined,
        "id": 16,
        "mainObjectId": undefined,
        "museumNo": "MusK24",
        "pathNames": undefined,
        "subNo": "b",
        "term": "Kokekar"
      },
      MusitObject {
        "breadcrumb": Array [
          Object {
            "id": 1,
            "name": "root-node",
            "url": "/magasin/1"
          },
          Object {
            "id": 2,
            "name": "Utviklingsmuseet",
            "url": "/magasin/2"
          },
          Object {
            "id": 3,
            "name": "Forskningens hus",
            "url": "/magasin/3"
          },
          Object {
            "id": 4,
            "name": "Kulturværelset",
            "url": "/magasin/4"
          },
          Object {
            "id": 5,
            "name": "Naturværelset",
            "url": "/magasin/5"
          }
        ],
        "currentLocationId": 5,
        "id": 9,
        "mainObjectId": undefined,
        "museumNo": "MusN28",
        "pathNames": Array [
          Object {
            "name": "root-node",
            "nodeId": 1
          },
          Object {
            "name": "Utviklingsmuseet",
            "nodeId": 2
          },
          Object {
            "name": "Forskningens hus",
            "nodeId": 3
          },
          Object {
            "name": "Kulturværelset",
            "nodeId": 4
          },
          Object {
            "name": "Naturværelset",
            "nodeId": 5
          }
        ],
        "subNo": undefined,
        "term": "Tusenben"
      }
    ],
    "totalMatches": 55
  },
  "loaded": true,
  "loading": false,
  "params": Object {
    "currentPage": 1,
    "perPage": 50
  }
}
`;
